[SaferCPP] Address issues in WebCore/html/canvas

roberto-apple opened the pull request:
#### b50bef42cf2f86954092779f1e2e20d879150de2
<pre>
[SaferCPP] Address issues in WebCore/html/canvas
<a href="https://rdar.apple.com/148673390">rdar://148673390</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291148">https://bugs.webkit.org/show_bug.cgi?id=291148</a>

Reviewed by Chris Dumez.

Address SaferCPP issues across multiple files in WebCore/html/canvas.

* Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::protectedCanvasBase const):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::validateBufferDataParameters):
(WebCore::WebGL2RenderingContext::validateBufferDataTarget):
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::addActiveContext):
(WebCore::WebGLRenderingContextBase::create):
(WebCore::WebGLRenderingContextBase::canvas):
(WebCore::WebGLRenderingContextBase::surfaceBufferToImageBuffer):
(WebCore::WebGLRenderingContextBase::validateBufferDataTarget):
(WebCore::WebGLRenderingContextBase::bufferData):
(WebCore::WebGLRenderingContextBase::bufferSubData):
(WebCore::WebGLRenderingContextBase::deleteRenderbuffer):
(WebCore::WebGLRenderingContextBase::deleteTexture):
(WebCore::WebGLRenderingContextBase::framebufferRenderbuffer):
(WebCore::WebGLRenderingContextBase::framebufferTexture2D):
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
(WebCore::WebGLRenderingContextBase::printToConsole):
(WebCore::WebGLRenderingContextBase::validateBufferDataParameters):
(WebCore::WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent):
(WebCore::WebGLRenderingContextBase::maybeRestoreContext):
(WebCore::WebGLRenderingContextBase::addDebugMessage):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/html/canvas/WebGLUtilities.cpp:
(WebCore::ScopedInspectorShaderProgramHighlight::shouldApply):
* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedWebGLRestoreFramebuffer::~ScopedWebGLRestoreFramebuffer):
* Source/WebCore/html/canvas/WebGLVertexArrayObject.h:
(isType):
* Source/WebCore/html/canvas/WebGLVertexArrayObjectBase.h:
* Source/WebCore/html/canvas/WebGLVertexArrayObjectOES.h:
(isType):

Canonical link: <a href="https://commits.webkit.org/293421@main">https://commits.webkit.org/293421@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a3553ffa4f5d415e8c24a9c41dfae485869c6d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18473 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103962 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49425 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100881 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18768 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26923 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75245 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32379 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101840 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14257 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89262 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55605 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14048 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7234 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48805 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83989 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7305 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106331 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18904 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84208 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83706 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21219 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28356 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6028 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19646 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25886 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31072 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25706 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27280 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -332,7 +332,7 @@ static void addActiveContext(WebGLRenderingContextBase& newContext)
     auto maxContextsSize = isMainThread() ? maxActiveContexts : maxActiveWorkerContexts;
     if (contexts.size() >= maxContextsSize) {
-        auto* earliest = *std::min_element(contexts.begin(), contexts.end(), [] (auto& a, auto& b) {
+        RefPtr<WebGLRenderingContextBase> earliest = *std::min_element(contexts.begin(), contexts.end(), [] (auto& a, auto& b) {
```
Please omit `<X>` (e.g. `<WebGLRenderingContextBase>` here) whenever possible everywhere in this PR. This should be `RefPtr earliest`

roberto-apple replied:
Removed explicit template arguments.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -758,7 +758,8 @@ bool WebGLRenderingContextBase::clearIfComposited(WebGLRenderingContextBase::Cal
 RefPtr<ImageBuffer> WebGLRenderingContextBase::surfaceBufferToImageBuffer(SurfaceBuffer sourceBuffer)
 {
-    auto buffer = canvasBase().buffer();
+    Ref base = canvasBase();
```
`canvasBase()` is only called once so we don't really need a local variable here. What we usually do is introduce a `protectedCanvasBase()` member function that returns a RefPtr and call it:
```
RefPtr buffer = protectedCanvasBase()->buffer();
```

roberto-apple replied:
Made requested changes for this and other comments below.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -978,7 +979,7 @@ bool WebGLRenderingContextBase::validateBufferTarget(ASCIILiteral functionName,
 WebGLBuffer* WebGLRenderingContextBase::validateBufferDataTarget(ASCIILiteral functionName, GCGLenum target)
 {
-    WebGLBuffer* buffer = nullptr;
+    RefPtr<WebGLBuffer> buffer = nullptr;
```
`= nullptr` is no longer necessary.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -994,7 +995,7 @@ WebGLBuffer* WebGLRenderingContextBase::validateBufferDataTarget(ASCIILiteral fu
         return nullptr;
     }
-    return buffer;
+    return buffer.get();
```
We should update this function to return a `RefPtr<>` given that this is what we have internally.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1165,8 +1166,8 @@ void WebGLRenderingContextBase::bufferData(GCGLenum target, std::optional<Buffer
 
-    std::visit([&](auto& data) {
-        m_context->bufferData(target, data->span(), usage);
+    std::visit([context = protectedGraphicsContextGL(), target, usage](auto& data) {
```
`context = m_context`. We don't normally use `protectedFoo()` functions to initialize variables.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1182,8 +1183,8 @@ void WebGLRenderingContextBase::bufferSubData(GCGLenum target, long long offset,
 
-    std::visit([&](auto& data) {
-        m_context->bufferSubData(target, static_cast<GCGLintptr>(offset), data->span());
+    std::visit([context = protectedGraphicsContextGL(), target, offset](auto& data) {
```
`context = m_context`

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -3149,8 +3148,11 @@ ExceptionOr<void> WebGLRenderingContextBase::texImageSourceHelper(TexImageFuncti
 
-    return std::visit([this, functionID, target, level, internalformat, border, format, type, xoffset, yoffset, zoffset, inputSourceImageRect, depth, unpackImageHeight](auto&& source) {
-        return texImageSource(functionID, target, level, internalformat, border, format, type, xoffset, yoffset, zoffset, inputSourceImageRect, depth, unpackImageHeight, *source);
+    return std::visit([weakThis = createRefForContextObject(), functionID, target, level, internalformat, border, format, type, xoffset, yoffset, zoffset, inputSourceImageRect, depth, unpackImageHeight](auto&& source) {
```
Just capture `this, protectedThis = Ref { *this }` here.

There is no need for a weak ptr, the lambda runs synchronously.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5018,22 +5020,22 @@ std::optional<std::span<const GCGLfloat>> WebGLRenderingContextBase::validateUni
 
 WebGLBuffer* WebGLRenderingContextBase::validateBufferDataParameters(ASCIILiteral functionName, GCGLenum target, GCGLenum usage)
```
We should update this function to return a RefPtr since this is what we have internally.

cdumez commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5167,17 +5169,15 @@ void WebGLRenderingContextBase::vertexAttribfvImpl(ASCIILiteral functionName, GC
-        if (isContextStopped())
-            return;
-        if (!isContextLost())
+    canvasBase().queueTaskKeepingObjectAlive(TaskSource::WebGL, [weakThis = createRefForContextObject()](auto&) {
```
Why not use the usual `= WeakPtr { *this }` here?

roberto-apple replied:
My initial thought process was that as I examined the class, the definition of createRefForContextObject() looked appropriate for a capture. I see now in Safer CPP Guidelines `weakThis = WeakPtr { *this }` is the suggested approach.

cdumez commented on Source/WebCore/html/canvas/WebGLVertexArrayObject.h:
```diff
@@ -44,8 +44,13 @@ class WebGLVertexArrayObject final : public WebGLVertexArrayObjectBase {
 
 } // namespace WebCore
 
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::WebGLVertexArrayObject) \
```
We don't need the `\` here and below.

cdumez commented on Source/WebCore/html/canvas/WebGLVertexArrayObjectOES.h:
```diff
@@ -39,8 +39,13 @@ class WebGLVertexArrayObjectOES final : public WebGLVertexArrayObjectBase {
 
 } // namespace WebCore
 
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::WebGLVertexArrayObjectOES) \
```
We don't need the `\` here and below.

cdumez reviewed (approved):
LGTM

roberto-apple committed:
[SaferCPP] Address issues in WebCore/html/canvas
rdar://148673390
https://bugs.webkit.org/show_bug.cgi?id=291148

Reviewed by Chris Dumez.

Address SaferCPP issues across multiple files in WebCore/html/canvas.

* Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::protectedCanvasBase const):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::validateBufferDataParameters):
(WebCore::WebGL2RenderingContext::validateBufferDataTarget):
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::addActiveContext):
(WebCore::WebGLRenderingContextBase::create):
(WebCore::WebGLRenderingContextBase::canvas):
(WebCore::WebGLRenderingContextBase::surfaceBufferToImageBuffer):
(WebCore::WebGLRenderingContextBase::validateBufferDataTarget):
(WebCore::WebGLRenderingContextBase::bufferData):
(WebCore::WebGLRenderingContextBase::bufferSubData):
(WebCore::WebGLRenderingContextBase::deleteRenderbuffer):
(WebCore::WebGLRenderingContextBase::deleteTexture):
(WebCore::WebGLRenderingContextBase::framebufferRenderbuffer):
(WebCore::WebGLRenderingContextBase::framebufferTexture2D):
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
(WebCore::WebGLRenderingContextBase::printToConsole):
(WebCore::WebGLRenderingContextBase::validateBufferDataParameters):
(WebCore::WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent):
(WebCore::WebGLRenderingContextBase::maybeRestoreContext):
(WebCore::WebGLRenderingContextBase::addDebugMessage):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/html/canvas/WebGLUtilities.cpp:
(WebCore::ScopedInspectorShaderProgramHighlight::shouldApply):
* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedWebGLRestoreFramebuffer::~ScopedWebGLRestoreFramebuffer):
* Source/WebCore/html/canvas/WebGLVertexArrayObject.h:
(isType):
* Source/WebCore/html/canvas/WebGLVertexArrayObjectBase.h:
* Source/WebCore/html/canvas/WebGLVertexArrayObjectOES.h:
(isType):

Canonical link: https://commits.webkit.org/293421@main
